compiler/rustc_mir_transform/src/ssa_range_prop.rs

```diff
@@ -0,0 +1,203 @@
+            if predecessors[bb].len() == 1 {
+                unique_predecessors.insert(bb);
+            }
+        }
+        RangeSet {
+            tcx,
+            typing_env,
+            ssa,
+            local_decls,
+            dominators,
+            ranges: FxHashMap::default(),
+            unique_predecessors,
+        }
+    }
+
+    /// Create a new known range at the location.
+    fn insert_range(&mut self, place: Place<'tcx>, location: Location, range: WrappingRange) {
+        self.ranges.entry(place).or_default().push((location, range));
+    }
+
+    /// Get the known range at the location.
+    fn get_range(&self, place: &Place<'tcx>, location: Location) -> Option<WrappingRange> {
+        let Some(ranges) = self.ranges.get(place) else {
+            return None;
+        };
+        // FIXME: This should use the intersection of all valid ranges.
+        let (_, range) =
+            ranges.iter().find(|(range_loc, _)| range_loc.dominates(location, &self.dominators))?;
+        Some(*range)
+    }
+
+    fn try_as_constant(
+        &mut self,
+        place: Place<'tcx>,
+        location: Location,
+    ) -> Option<ConstOperand<'tcx>> {
+        if let Some(range) = self.get_range(&place, location)
+            && range.start == range.end
+        {
+            let ty = place.ty(self.local_decls, self.tcx).ty;
+            let layout = self.tcx.layout_of(self.typing_env.as_query_input(ty)).ok()?;
+            let value = ConstValue::Scalar(Scalar::from_uint(range.start, layout.size));
+            let const_ = Const::Val(value, ty);
+            return Some(ConstOperand { span: DUMMY_SP, user_ty: None, const_ });
+        }
+        None
+    }
+
+    fn simplify_operand(
+        &mut self,
+        operand: &mut Operand<'tcx>,
+        location: Location,
+    ) -> Result<(), Option<Place<'tcx>>> {
+        let Some(place) = operand.place() else {
+            return Ok(());
+        };
+        let Some(const_) = self.try_as_constant(place, location) else {
+            if self.is_ssa(place) {
+                return Err(Some(place));
+            } else {
+                return Err(None);
+            }
+        };
+        *operand = Operand::Constant(Box::new(const_));
+        Ok(())
+    }
+
+    fn is_ssa(&self, place: Place<'tcx>) -> bool {
+        self.ssa.is_ssa(place.local) && place.is_stable_offset()
+    }
+}
+
+impl<'tcx> MutVisitor<'tcx> for RangeSet<'tcx, '_, '_> {
+    fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+
+    fn visit_operand(&mut self, operand: &mut Operand<'tcx>, location: Location) {
+        let _ = self.simplify_operand(operand, location);
+    }
+
+    fn visit_terminator(&mut self, terminator: &mut Terminator<'tcx>, location: Location) {
+        match &mut terminator.kind {
+            TerminatorKind::Assert { cond, expected, target, .. } => {
+                if let Err(Some(place)) = self.simplify_operand(cond, location) {
```

Comment:
Can this be simplified to:
```suggestion
                self.simplify_operand(cond, location);
                if let Some(place) = cond.place() {
```